include/zephyr/math/util.h

```diff
@@ -0,0 +1,110 @@
+/**
+ * @file math/util.h
+ *
+ * @brief A common math library for use in both floating and fixed point.
+ */
+
+/*
+ * Copyright (c) 2022 Google Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef INCLUDE_MATH_UTIL_H_
+#define INCLUDE_MATH_UTIL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef CONFIG_FPU
+#define PRIf "f"
+#define PRIf_ARG(n) (n)
+#else
+#define PRIf "d.%06llu"
+#define __FP_DECIMAL(n) ((n) & ((1 << FP_BITS) - 1))
+#define PRIf_ARG(n)                                                                                \
+	FP_TO_INT(n),                                                                              \
+	((__FP_DECIMAL(n) * UINT64_C(1000000)) + ((1 << FP_BITS) / 2)) / (1 << FP_BITS)
+#endif
+
+#ifdef CONFIG_FPU
+typedef float fp_t;
+typedef float fp_inter_t;
+/* Conversion to/from fixed-point */
+#define INT_TO_FP(x) ((float)(x))
+#define FP_TO_INT(x) ((int32_t)(x))
+/* Float to fixed-point, only for compile-time constants and unit tests */
+#define FLOAT_TO_FP(x) ((float)(x))
+/* Fixed-point to float, for unit tests */
+#define FP_TO_FLOAT(x) ((float)(x))
+
+#define FLT_MAX (3.4028234664e+38)
+#define FLT_MIN (1.1754943508e-38)
+
+#else
+/* Fixed-point type */
+typedef int32_t fp_t;
+
+/* Type used during fp operation */
+typedef int64_t fp_inter_t;
+
+/* Number of bits left of decimal point for fixed-point */
+#define FP_BITS 16
```

Comment:
We can't assume that all fixed-point quantities will use the same number of fractional bits.

Reply:
I just want to make sure that I understand correclty. The old API uses a `struct sensor_value` which uses an `int32_t` for the integer part and the same for the fractional part. This effectively does the same thing. Or is the concern that I'm just using 32 bits in total vs the original 64 bits? I'm happy to tweak this, for now it's just the same accuracy that we use on Chrome so I didn't want to sink a lot of time into this until we have better reviewer the API.

Reply:
I dislike how the current sensor API defines integer and fractional parts (particularly that the fractional part is decimal, which makes for some awkward arithmetic). The ideal number of fixed-point fractional bits is an application-specific tradeoff of range vs. precision, and depends on the choice of units for a given type of sensor.

Reply:
Would a Kconfig here work?

Reply:
@yperess @MaureenHelm It would be great if new api is compatible with what is already being used. Zephyr already has a legacy that it simply must take into account. If previous precision was u32 + u32 there should be an option to still use it.
I also recommend to not switch to float type solely by CONFIG_FPU. It might be that integrator would still prefer to use integers for sensor api (for memory or compatibility reasons).

Reply:
My five cents here: I'd expect a path forward for all in-tree sensors. I'd not introduce a new “V2” API without a proper migration plan, considering the number of sensors we have. I'd also like to see more sensors ported (different sensor types, preferably) to see if the API is capable of adapting to all (or most) sensors we have.